### PR TITLE
[bug] - Ensure BufferedFileWriter Flushes Buffer Contents to File Correctly

### LINKED ISSUE #2927: Valid zip file: `error extracting archive with format: .zip: zip: not a valid zip file`
Please review the [Community Note](https://github.com/trufflesecurity/trufflehog/blob/main/.github/community_note.md) before submitting

### TruffleHog Version
https://github.com/trufflesecurity/trufflehog/commit/b0fd70c0ffb7dd9e4dc2fb3e844f7fdadf332e9e

### Trace Output
```
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷
...
2024-06-05T22:01:35-04:00       info-2  trufflehog      enumerating source      {"source_manager_worker_id": "QAtyb"}
2024-06-05T22:01:35-04:00       info-3  trufflehog      chunking unit   {"source_manager_worker_id": "QAtyb", "unit": "/tmp/whitesource-fs-agent-18.1.1.jar", "unit_kind": "unit"}
2024-06-05T22:01:35-04:00       info-3  trufflehog      scanning file   {"source_manager_worker_id": "QAtyb", "unit": "/tmp/whitesource-fs-agent-18.1.1.jar", "unit_kind": "unit", "path": "/tmp/whitesource-fs-agent-18.1.1.jar"}
2024-06-05T22:01:35-04:00       error   trufflehog      error unarchiving chunk.        {"source_manager_worker_id": "QAtyb", "unit": "/tmp/whitesource-fs-agent-18.1.1.jar", "unit_kind": "unit", "timeout": 30, "error": "error extracting archive with format: .zip: zip: not a valid zip file"}
2024-06-05T22:01:35-04:00       info-5  trufflehog      handler channel closed, all chunks processed    {"source_manager_worker_id": "QAtyb", "unit": "/tmp/whitesource-fs-agent-18.1.1.jar", "unit_kind": "unit"}
```

### Expected Behavior

Valid zip files should be extracted.

### Actual Behavior

TruffleHog is saying there's an error with a [valid zip file](https://github.com/whitesource/fs-agent/blob/a38d313bb2f5311b831fa7252d9812ab955459cd/src/standAlone/whitesource-fs-agent-18.1.1.jar) and not scanning it.

```sh
$ Archive:  whitesource-fs-agent-18.1.1.jar
   creating: META-INF/
  inflating: META-INF/MANIFEST.MF
...
$ echo $?
0
```

### Steps to Reproduce

 1. Download https://github.com/whitesource/fs-agent/blob/a38d313bb2f5311b831fa7252d9812ab955459cd/src/standAlone/whitesource-fs-agent-18.1.1.jar
 2. Run `./trufflehog filesystem .whitesource-fs-agent-18.1.1.jar`
 3. Observe the reported error

## Environment
```sh
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

## Additional Context
N/A

### References
N/A

### LINKED ISSUE #2932: Valid `.rpm` file: `error reading RPM payload: xz: data is corrupt`
Please review the [Community Note](https://github.com/trufflesecurity/trufflehog/blob/main/.github/community_note.md) before submitting

### TruffleHog Version
3.78.0

### Trace Output
```
2024-06-05T23:13:18-04:00       error   trufflehog      error processing RPM files      {"source_manager_worker_id": "Pe0Ht", "repo": "https://github.com/openshift/origin-web-console-smoke-test.git", "commit": "a6a22f8", "path": "google-chrome-stable-59.rpm", "timeout": 30, "error": "error reading RPM payload: xz: data is corrupt"}
```

https://github.com/trufflesecurity/trufflehog/blob/467c4232c9644e5895d6042ad88caef777d46356/pkg/handlers/rpm.go#L81

### Expected Behavior

The RPM file should be extracted and scanned.

### Actual Behavior

The RPM file is not extracted due to an error.

```sh
$ rpm2cpio google-chrome-stable-59.rpm  | cpio -idmv
./etc/cron.daily/google-chrome
./opt/google/chrome
./opt/google/chrome/chrome
...
./usr/share/man/man1/google-chrome.1
466184 blocks
```

### Steps to Reproduce

 1. Download https://github.com/openshift/origin-web-console-smoke-test/blob/a6a22f8/google-chrome-stable-59.rpm
 2. Run trufflehog
 3. Observe reported error

## Environment
 N/A

## Additional Context
N/A

### References
N/A

### LINKED ISSUE #2936: Valid `.tgz` archive: `corrupt input before offset 10562302`
Please review the [Community Note](https://github.com/trufflesecurity/trufflehog/blob/main/.github/community_note.md) before submitting

### TruffleHog Version
3.78.0

### Trace Output
```
./trufflehog filesystem /tmp/nr-firehose-nozzle-0.0.3.tgz
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷

2024-06-06T19:06:46-04:00       info-0  trufflehog      running source  {"source_manager_worker_id": "iVlJB", "with_units": true}
2024-06-06T19:06:46-04:00       error   trufflehog      error unarchiving chunk.        {"source_manager_worker_id": "iVlJB", "unit": "/tmp/nr-firehose-nozzle-0.0.3.tgz", "unit_kind": "unit", "timeout": 30, "error": "error creating custom reader: error creating random access reader: error creating BufferedFileReader: error writing to buffered file writer: flate: corrupt input before offset 10562302"}
2024-06-06T19:06:46-04:00       info-0  trufflehog      finished scanning       {"chunks": 0, "bytes": 0, "verified_secrets": 0, "unverified_secrets": 0, "scan_duration": "44.490834ms", "trufflehog_version": "dev"}
```

### Expected Behavior

A valid `.tgz` should be extracted and scanned.

```sh
$ tar -tvf nr-firehose-nozzle-0.0.3.tgz
-rw-r--r-- shahram/staff   860 2017-08-27 23:41 ./release.MF
drwxr-xr-x shahram/staff     0 2017-08-27 23:41 ./jobs/
-rw-r--r-- shahram/staff  2434 2017-08-27 23:41 ./jobs/delete-all.tgz
-rw-r--r-- shahram/staff  4533 2017-08-27 23:41 ./jobs/deploy-all.tgz
drwxr-xr-x shahram/staff     0 2017-08-27 23:41 ./packages/
-rw-r--r-- shahram/staff 5841596 2017-08-27 23:41 ./packages/cf_cli.tgz
-rw-r--r-- shahram/staff 25649292 2017-08-27 23:41 ./packages/newrelic_firehose_nozzle.tgz
$ tar xzf nr-firehose-nozzle-0.0.3.tgz && echo $?
0
```

### Actual Behavior

An error occurs and the archive isn't scanned

### Steps to Reproduce

1. Download https://github.com/newrelic/newrelic-pcf-nozzle-tile/blob/0cc98a162d1b36fead8d5d725c9126a1ffae07a3/releases/nr-firehose-nozzle-0.0.3.tgz
2. Run trufflehog against the file
3. Observe reported error

## Environment
N/A

## Additional Context
N/A

### References
N/A

### PR DESCRIPTION
**Pull request by ahrav**
### Description:
**Problem:**
The `BufferedFileWriter` was not correctly flushing the buffer contents to the file once the buffer threshold was exceeded and a file was created. This caused subsequent writes that exceeded the threshold to bypass the buffer without flushing it, leading to incorrect data in the file.

**Fix:**
Ensure that any remaining data in the buffer is always flushed to the file once a file is created, before writing new data. This guarantees that the file contains the correct data.

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

File: pkg/writers/buffered_file_writer/bufferedfilewriter.go [hunk 1]
```diff
@@ -205,7 +205,13 @@ func (w *BufferedFileWriter) Write(data []byte) (int, error) {
 				}
 				return 0, err
 			}
-			w.buf.Reset()
```
Comment (ahrav) [inline pkg/writers/buffered_file_writer/bufferedfilewriter.go:208]: This was redundant. `WriteTo` already resets the buffer.

**Comment by rgmz** [comment]
Based on my testing:

- [x] Fixes #2927
- [ ] Does not fix #2928
- [x] Fixes #2932
- [x] Fixes #2936

**Comment by rgmz** [comment]
Likewise. :) 

It's probable that #2928 has a different, albeit similar, cause.

**Comment by rgmz** [comment]
> It's probable that #2928 has a different, albeit similar, cause.

farnsworth

I've determined that the issue is with the upstream `archiver` library. It's actually already been fixed, but the fix hasn't been released yet.

**Comment by ahrav** [comment]
> > Does not fix #2928
> 
> farnsworth
> 
> I've determined that the issue is with the upstream `archiver` library. It's actually already been fixed, but the fix hasn't been released yet.

i'll gladly take a freebie any day 🥳

**Review by rosecodym — APPROVED** [review]

**Pull request merged**